packages/ai/src/core.ts

```diff
@@ -68,37 +69,68 @@ export class EdgeDBAI {
     });
   }
 
-  private async fetchRag(request: Omit<RAGRequest, "model" | "prompt">) {
+  private async fetchRag(request: RagRequest, context: QueryContext) {
     const headers = request.stream
       ? { Accept: "text/event-stream", "Content-Type": "application/json" }
       : { Accept: "application/json", "Content-Type": "application/json" };
 
+    const { messages: initialMessages, prompt } = request;
+
+    if (prompt && initialMessages)
+      throw new Error(
+        "You can provide either a prompt or a messages array, not both.",
+      );
+
+    const messages: EdgeDBMessage[] | undefined = prompt
+      ? [{ role: "user" as const, content: [{ type: "text", text: prompt }] }]
+      : initialMessages;
+
+    const providedPrompt =
+      this.options.prompt &&
+      ("name" in this.options.prompt || "id" in this.options.prompt);
+
     const response = await (
       await this.authenticatedFetch
     )("rag", {
       method: "POST",
       headers,
       body: JSON.stringify({
         ...request,
+        context,
         model: this.options.model,
-        prompt: this.options.prompt,
+        prompt: {
+          ...this.options.prompt,
+          // if user provides prompt.custom without id/name it is his choice
+          // to not include default prompt msgs, but if user provides messages
+          // and doesn't provide prompt.custom, since we add messages to the
+          // prompt.custom we also have to include default prompt messages
+          ...(!this.options.prompt?.custom &&
+            !providedPrompt && {
+              name: "builtin::rag-default",
+            }),
+          custom: [...(this.options.prompt?.custom || []), ...messages!],
+        },
+        query: [...messages!].reverse().find((msg) => msg.role === "user")!
+          .content[0].text,
       }),
     });
 
     if (!response.ok) {
       const bodyText = await response.text();
-      throw new Error(bodyText);
+      throw new Error(JSON.parse(bodyText).message);
```

Comment:
I'd be careful about parsing the arbitrary text here. It'll be JSON when we return an error, but if we hit some weird case (like CDN returning a 503 or something) that returns HTML or plain text, this will throw a totally unhelpful error.

If you want to make it a nicer message when it's JSON, I'd say do something like: check the content-type of the response, and if it's JSON-like, use `await response.json()`, otherwise use just dump the full text into the error.

Reply:
I believe u were thinking sth like this:
```
    if (!response.ok) {
      const contentType = response.headers.get("content-type");
      let errorMessage: string;

      if (contentType && contentType.includes("application/json")) {
        const json = await response.json();
        errorMessage = json.message || "An error occurred";
      } else {
        const bodyText = await response.text();
        errorMessage = bodyText || "An unknown error occurred";
      }
      throw new Error(errorMessage);
    }
```

Reply:
Yeah, something like that! Although, you might want to do something a bit more strict for JSON since JSON could be all kinds of shapes (including a `null` literal), like:

```ts
        if (typeof json === "object" && json != null && "message" in json) {
          errorMessage = json.message;
        } else {
          errorMessage = `An error occurred: ${json}`;
        }
```

Definitely a lot more work than returning the unparsed text, and probably can/should move into some general purpose error handling util function. I just want to avoid a runtime error that will make it hard to tell what happened.